doc/CONFIGURATION.md

```diff
@@ -227,6 +228,27 @@ ExecStop=/usr/bin/fusermount -u /home/ec2-user/s3-bucket-mount
 WantedBy=remote-fs.target
 ```
 
+## Caching configuration
+
+Mountpoint can optionally cache object metadata and content to reduce cost and improve performance for repeated reads to the same file.
+
+To enable caching, use the `--cache <CACHE_DIR>` command-line flag.
+This flag will enable caching of metadata using a default time-to-live (TTL) of 60 minutes.
+Object/file content will also be cached within the cache directory specified.
+Mountpoint will create a directory within this path, removing any files or directories within it at mount time and exit.
```

Comment:
Don't think we should be deleting at exit time — that's something we can't change our mind about if we want to default to reusable cache later.

Reply:
I'm not sure what to pick here. Two paths forward.

- We don't do cleanup at mount or exit.
  - There is a risk that we fill up the disk with stale cache content that we don't track, since we don't attempt to ingest old cache data. We eventually get to a point where we have very little space remaining for the real cache data, unless the customer takes manual action.
  - We also provide a `--delete-cache` option which will delete the cache at mount and exit.
  - Note, we haven't defined our recommendations to the user. We haven't tested disk reuse, and I assume we should discourage it here in the documentation until we are happy that cache reuse is both well tested and supported.
- We do cleanup at mount and exit.
  - We mitigate the risk from the disk filling up.
  - We don't need to guide the customer on cache reuse, since its not possible to use it.
  - Allowing cache reuse later would potentially be a breaking change since we'd change the disk usage characteristics, and would likely need to be opt-in through some `--allow-cache-reuse`/`--no-delete-cache` flag.

I do think I actually prefer the former option, but we need to guide users in the right direction to ensure they aren't reusing the cache.

Reply:
My other thinking is around how we should cleanup the cache.

At the moment, we are planning to delete the `mp-cache-V1` directory. I think this is fine, but in this case I think we also need to guide users to using a unique value for `--cache <CACHE_DIR>` for different buckets so that the caches don't trample over each other. The other option is that we include a hash of the bucket name into the path (which I think we should do anyway) and _delete only that sub-directory, leaving any other buckets intact_. I don't see much issue with this either actually.

Reply:
Discussed on a call with @jamesbornholt and @passaro, documenting here:

- We'll implement the cache directory argument as a directory we are creating a sub-directory within, and deleting the contents at mount and at exit. This avoids issues where requiring a non-empty directory causes availability issues on unexpected process restarts, or with not deleting and disk space accumulating and not being cleaned.
- Configuration documentation will recommend to use a unique value for this argument when running multiple processes, to avoid any clobbering of each process' cache. If Mountpoint is used with the same path, it may cause performance issues when the cache is clobbered but we also expect mounting and un-mounting on a process to be relatively infrequent and the impact here should be limited.
- If we were to consider cache re-use in the future, its unlikely we can silently change behavior to allow re-use and we'd probably need a new argument for that anyway.